cpp/ql/src/Likely Bugs/UnderspecifiedFunctions/MistypedFunctionArguments.ql

```diff
@@ -0,0 +1,117 @@
+  or
+  arg instanceof CharType and parm instanceof LongLongType
+  or
+  arg instanceof ShortType and parm instanceof IntType
+  or
+  arg instanceof ShortType and parm instanceof LongType
+  or
+  arg instanceof ShortType and parm instanceof LongLongType
+  or
+  arg instanceof IntType and parm instanceof LongType
+  or
+  arg instanceof IntType and parm instanceof LongLongType
+  or
+  arg instanceof LongType and parm instanceof LongLongType
+  or
+  arg instanceof Enum and parm instanceof IntType
+  or
+  arg instanceof Enum and parm instanceof LongType
+  or
+  arg instanceof Enum and parm instanceof LongLongType
+  or
+  // enums are usually sized as ints
+  arg instanceof IntType and parm instanceof Enum
+  or
+  // pointer types
+  arg.(PointerType).getBaseType().getUnspecifiedType() = parm
+        .getUnspecifiedType()
+        .(PointerType)
+        .getBaseType()
+        .getUnspecifiedType()
+  or
+  // void * parameters accept arbitrary pointer arguments
+  arg instanceof PointerType and
+  parm.(PointerType).getBaseType().getUnspecifiedType() instanceof VoidType
+  or
+  // handle reference types
+  argTypeMayBePromoted(arg.(ReferenceType).getBaseType().getUnspecifiedType(), parm)
+  or
+  argTypeMayBePromoted(arg, parm.(ReferenceType).getBaseType().getUnspecifiedType())
+  or
+  // array-to-pointer decay
+  argTypeMayBePromoted(arg.(ArrayType).getBaseType().getUnspecifiedType(),
+    parm.(PointerType).getBaseType().getUnspecifiedType())
+  or
+  // pointer-to-array promotion (C99)
+  argTypeMayBePromoted(arg.(PointerType).getBaseType().getUnspecifiedType(),
+    parm.(ArrayType).getBaseType().getUnspecifiedType())
+}
+
+// This predicate doesn't necessarily have to exist, but if it does exist
+// then it must be inline to make sure we don't enumerate all pairs of
+// compatible types.
+// Its body could also just be hand-inlined where it's used.
+pragma[inline]
+predicate argMayBePromoted(Expr arg, Parameter parm) {
+  argTypeMayBePromoted(arg.getExplicitlyConverted().getType().getUnspecifiedType(),
+    parm.getType().getUnspecifiedType())
+  or
+  // The value 0 is often passed in to indicate NULL, or to initialize an arbitrary integer type.
+  // we will allow all literal values for simplicity.  Pointer parameters are sometime passed
+  // special-case literals such as -1, -2, etc.
+  arg.(Literal).getType() instanceof IntegralType and
+  (
+    parm.getType().getUnspecifiedType() instanceof PointerType
+    or
+    parm.getType().getUnspecifiedType() instanceof IntegralType and
+    arg.(Literal).getType().getSize() <= parm.getType().getUnspecifiedType().getSize()
+  )
+}
+
+from FunctionCall fc, Function f, Parameter p
+where
+  f = fc.getTarget() and
+  p = f.getAParameter() and
+  not f.isVarargs() and
+  p.getIndex() < fc.getNumberOfArguments() and
+  // There must be a zero-parameter declaration
+  exists(FunctionDeclarationEntry fde | fde = f.getADeclarationEntry() |
+    fde.getNumberOfParameters() = 0
+  ) and
```

Comment:
The [FPs in numpy](https://lgtm.com/query/3622830835849000749/) are interesting. They arise because a careless `configure` test has declared `sqrt` with no parameters, and unfortunately that test got extracted. I'm thinking about how to avoid this in the autobuilder, but it's not easy.

Even without that autobuilder quirk, a false positive like this could still occur if the function is declared without prototype somewhere in the project and then used with a prototype somewhere else, unless our linker awareness is able to disambiguate it. I think we could fix that by adding a library predicate that links a call site to the `FunctionDeclarationEntry`s that were in scope at that call. Then we should only consider a call to have a zero-parameter declaration if there was such a declaration _in scope_.

See the point below about `getFullyConverted`, though. It may fix all cases of this problem in a simpler way. When that's fixed, I wonder whether we need this check for "There must be a zero-parameter declaration" at all. It probably suppresses many FPs in practice, but could we craft code with each of those FPs that gets around this check?